### PR TITLE
Docker development stack docs

### LINKED ISSUE #251: More detailed documentation on drunc and Docker stack
Additional details to document:
- How and where different procesess run.
- SSH server running within drunc image.
- Dark magic of Dockerfile.
- Source of base image and procedure to update.
- Drunc dependencies.
- Dune development environment.

### PR DESCRIPTION
**Pull request by cc-a**
# Description

Provides more detailed documentation on the dependencies of the project and how these are implemented in the our Docker development stack.

Fixes #251 

## Type of change

- [X] Documentation (non-breaking change that adds or improves the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (non-breaking, back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (whatever its nature)

## Key checklist

- [ ] All tests pass (eg. `python -m pytest`)
- [X] The documentation builds and looks OK (eg. `python -m sphinx -b html docs docs/build`)
- [X] Pre-commit hooks run successfully (eg. `pre-commit run --all-files`)

## Further checks

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added or an issue has been opened to tackle that in the future. (Indicate issue here: # (issue))

### CODE REVIEW

**Review by jamesturner246 — APPROVED** [review]
Looks good. A few tiny things.

File: docs/development.md [hunk 1]
```diff
@@ -55,3 +55,131 @@ Pre-commit should now work as expected when making commits even without the need
...
+
+The base images are published by the [daq-release repository]. There are separate images
+for nightly ([nightly-release-alma9]) and release ([frozen-release-alma9]) builds. We
+are currently tracking nightly builds as we are dependent on cutting edge drunc
+developments. It may be feasible in the future to swap to a pinned release build.
+
+To ensure a consistent developer experience we pin the development image to a fixed base
+image. The base image is manually tagged and pushed to the
+[imperialcollegelondon/dunedaq_dev_environment] package with a tag including the date of
+the nightly.
```
Comment (dalonsoa) [inline docs/development.md:108]: Why cannot we directly point to the chosen nightly build in our dockerfile and instead have to download the base image, rename and push again?

File: docs/development.md [hunk 2]
```diff
@@ -55,3 +55,131 @@ Pre-commit should now work as expected when making commits even without the need
...
+launches processes on localhost via SSH. To support this the `drunc` service also starts
+an SSH server with keys configured to allow passwordless connection to localhost.
+
+#### Updating the drunc base image
+
+The following changes are needed to update the version of the base image used for the
+`drunc` service.
+
+1. Pull the latest nightly image - `docker pull
+   ghcr.io/dune-daq/nightly-release-alma9:development_v5`.
```
[the remark this answers is not in the document]
Comment (dalonsoa) [reply docs/development.md:121]: I think we want always to choose a specific version, and not just the latest, to have better control of what we are actually using.

File: docs/development.md [hunk 3]
```diff
@@ -55,3 +55,131 @@ Pre-commit should now work as expected when making commits even without the need
...
+1. Retag the image for the imperialcollegelondon/dunedaq_dev_environment package using
+   the date of the new nightly, e.g.
+   `docker tag ghcr.io/dune-daq/nightly-release-alma9:development_v5
+   ghcr.io/imperialcollegelondon/dunedaq_dev_environment:nightly-241114`.
+1. Update `drunc_docker_service/Dockerfile` to the reflect the date of the new nightly.
+   You'll need to change the `NIGHTLY_TAG` environment variable and the name of the base
+   image.
```
Comment (dalonsoa) [inline docs/development.md:128]: Maybe mention that only the date part of the `NIGHTLY_TAG` should be updated, so `NFD_DEV_241114_A9` becomes `NFD_DEV_NEWDATE_A9`, eg. `NFD_DEV_241206_A9`.

File: docs/development.md [hunk 4]
```diff
@@ -55,3 +55,131 @@ Pre-commit should now work as expected when making commits even without the need
...
+1. Update the pinned commits of `drunc` and `druncschema` in `pyproject.toml` to match
+   the versions in the image. This is usually the latest commit in the `develop` branch
+   of those repositories but you should check that no new commits have been added since
+   the nightly was built.
+1. Regenerate the Poetry lock file.
```
Comment (dalonsoa) [inline docs/development.md:133]:
```suggestion
1. Regenerate the Poetry lock file with `poetry lock`.
```
Possibly, `poetry lock --no-update`? Not sure.

**Review by dalonsoa — APPROVED** [review]
I've just made a couple of suggestions, but ti looks good, otherwise.

File: docs/development.md (hunk 1 above)
Comment (cc-a) [reply docs/development.md:108]: The nightly images are not retained by the dune project so we have to take our own copy and point to that.

File: docs/development.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (cc-a) [reply docs/development.md:121]: The `development_v5` is the relevant rolling tag for the nightly images.

File: docs/development.md (hunk 3 above)
Comment (cc-a) [reply docs/development.md]: Done

**Pull request merged**